apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Filter.cs

```diff
@@ -0,0 +1,240 @@
+        }
+
+        private static UnaryOp GetUnaryOperator(object value, FieldOp op)
+        {
+            switch (value)
+            {
+                case null:
+                    return op switch
+                    {
+                        FieldOp.Equal => UnaryOp.IsNull,
+                        FieldOp.NotEqual => UnaryOp.IsNotNull,
+                        _ => throw new ArgumentException("Null values can only be used with the Equal/NotEqual operators", nameof(value))
+                    };
+                case double d when double.IsNaN(d):
+                case float f when float.IsNaN(f):
+                    return op switch
+                    {
+                        FieldOp.Equal => UnaryOp.IsNan,
+                        FieldOp.NotEqual => UnaryOp.IsNotNan,
+                        _ => throw new ArgumentException("Not-a-number values can only be used with the Equal/NotEqual operators", nameof(value))
+                    };
+                default:
+                    return UnaryOp.Unspecified;
+            }
+        }
+
+        public bool Equals(UnaryFilter other) => throw new NotImplementedException();
+    }
+
+    /// <summary>
+    /// 
+    /// </summary>
+    internal class CompositeFilter : Filter, IEquatable<CompositeFilter>
+    {
+        /// <summary>
+        /// 
+        /// </summary>
+        public IEnumerable<Filter> Filters { get; }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        public Operator Operator { get; }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="filters"></param>
+        /// <param name="op"></param>
+        public CompositeFilter(IEnumerable<Filter> filters, Operator op)
+        {
+            Filters = new List<Filter>(filters);
+            Operator = op;
+        }
+
+        public CompositeFilter WithAdditionalFilter(Filter filter)
+        {
+            var filters = new List<Filter>(Filters);
+            filters.Add(filter);
+            return new CompositeFilter(filters, Operator);
+        }
+
+        internal override ProtoFilter ToProto(SerializationContext context) =>
+            new ProtoFilter { CompositeFilter = new StructuredQuery.Types.CompositeFilter { Op = Operator, Filters = { Filters.Select(f => f.ToProto(context)) } } };
+
+        
+
+        public bool Equals(CompositeFilter other) => throw new NotImplementedException();
+    }
+
+    internal static bool IsOrderingFilter(UnaryFilter filter) =>
+        filter.Value is object &&
+        filter.Op == (int) FieldOp.GreaterThan ||
+        filter.Op == (int) FieldOp.GreaterThanOrEqual ||
+        filter.Op == (int) FieldOp.LessThan ||
+        filter.Op == (int) FieldOp.LessThanOrEqual;
+
+    internal static UnaryFilter GetOrderingFilter(CompositeFilter compositeFilter)
+    {
+        foreach (var filter in compositeFilter.Filters)

```

Comment:
@jskeet Want your feedback on this ordering part. We are doing DFS to find the first ordering filter and using it as default one. Please share your thoughts on this.

Reply:
That's not something I can answer - please ask the Firestore SDK team what their requirements are.

Reply:
Thanks Jon will check with them. Thanks for the feedback. I will raise the final PR now with all the changes.